src/command/expire/expire.c

```diff
@@ -863,6 +863,88 @@ removeExpiredBackup(InfoBackup *infoBackup, const String *adhocBackupLabel, unsi
     FUNCTION_LOG_RETURN_VOID();
 }
 
+
+/***********************************************************************************************************************************
+Remove expired backup history manifests from repo
+***********************************************************************************************************************************/
+static void
+removeExpiredHistory(InfoBackup *infoBackup, unsigned int repoIdx)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(INFO_BACKUP, infoBackup);
+        FUNCTION_LOG_PARAM(UINT, repoIdx);
+    FUNCTION_LOG_END();
+
+    ASSERT(infoBackup != NULL);
+
+    MEM_CONTEXT_TEMP_BEGIN()
+    {
+        if (cfgOptionIdxTest(cfgOptRepoRetentionHistory, repoIdx))
+        {
+            // Get all the current backups in backup.info - these will not be expired.
+            StringList *currentBackupList = strLstSort(infoBackupDataLabelList(infoBackup, NULL), sortOrderDesc);
+
+            // If the full backup history manifests are expired, we should expire diff and incr too. So, format the oldest full
+            // backup label to retain.
+            const time_t minTimestamp = time(NULL) - (time_t)(cfgOptionIdxUInt(cfgOptRepoRetentionHistory, repoIdx) * SEC_PER_DAY);
+            char buffer[16];
+            THROW_ON_SYS_ERROR(strftime(
+                buffer, sizeof(buffer), "%Y%m%d-%H%M%S", localtime(&minTimestamp)) == 0, AssertError, "unable to format time");
+            String *minBackupLabel = strNewFmt("%sF", buffer);
```

Comment:
There is already a function in `command/backup/backup.c` to do this formatting, `backupLabelFormat()`. I think it would be best to move that function into `command/backup/common.c` and reuse it. You may also need to move some tests to get coverage.

Reply:
Agreed. I first wanted to try that but wondered if it should be part of this patch or a new one

Reply:
Since this PR is what is driving the need to have `backupLabelFormat()` more available, I'd suggest the move to `common.c` be part of this PR.

Reply:
Is there any precedence in the include files? I'm struggling with an awkward error.

Moving `backupLabelFormat()` to `common.c`, I also have to include `infoBackup.h` to support **BackupType**. But then, I get a compilation error:
```
In file included from ./command/backup/common.h:11:0,
                 from ./info/manifest.h:13,
                 from info/manifest.c:17:
./info/infoBackup.h:126:1: error: unknown type name ‘Manifest’
 void infoBackupDataAdd(const InfoBackup *this, const Manifest *manifest);
 ^
cc1: warning: unrecognized command line option "-Wno-implicit-fallthrough" [enabled by default]
make: *** [.build/info/manifest.o] Error 1
```

Would it rather make sense to move the `BackupType` definition to `common.h`?

Reply:
> Would it rather make sense to move the BackupType definition to common.h?

No, that will definitely cause a lot of breakage. The issue was the ordering of the header files in manifest.h, see dc0e6645cd5c4.

Pull from master and you should be good to go. Please don't rebase as we'll lose the PR history.

Reply:
fetch upstream + merge upstream/master solved it indeed. Thanks !
I'll look at the tests and push afterwards then :-)

Reply:
Just pushed 520da96.
`--vm=u20 --c-only` tests are running fine on my laptop. I'll watch the CI to check if more changes are needed in the tests.
Thanks again for your help!

Reply:
Yes, this looks pretty good, but also be sure to use this function to simply formatting in the tests.

Reply:
@cmwshang It's a bit hard to migrate the tests here since this function is called through the parent function in the tests. Perhaps this is the time to combine the tests? It would make moving functions around less painful.

Reply:
Tests updated in d70c052. Including `common.h` was enough apparently.